src/attr.rs

```diff
@@ -237,6 +246,74 @@ impl<'s> std::fmt::Debug for Attributes<'s> {
     }
 }
 
+/// Iterator over [Attributes] key-value pairs, in arbitrary order.
+pub struct AttributeIntoIter<'s> {
+    items: VecDeque<(&'s str, AttributeValue<'s>)>,
+}
+
+impl<'s> Iterator for AttributeIntoIter<'s> {
+    type Item = (&'s str, AttributeValue<'s>);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.items.pop_front()
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let len = self.items.len();
+        (len, Some(len))
+    }
+}
+
+impl<'s> IntoIterator for Attributes<'s> {
+    type Item = (&'s str, AttributeValue<'s>);
+
+    type IntoIter = AttributeIntoIter<'s>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        AttributeIntoIter {
+            items: self.0.map_or(vec![].into(), |b| (*b).into()),
+        }
+    }
+}
+
+/// Iterator over references to [Attributes] key-value pairs, in arbitrary order.
+pub struct AttributeIter<'s> {
+    attrs: Option<&'s [(&'s str, AttributeValue<'s>)]>,
+    idx: usize,
```

Comment:
should be possible with just a single slice and no index/option also, right? start of slice moves forward on next and None is an empty slice.

could even just make AttributesIter into a newtype for std::slice::Iter? I suspect that implementation is well thought out.

I guess AttributeIntoIter could be a newtype of std::vec::IntoIter as well.

(newtypes instead of type aliases in order to hide the internal implemenation and allow it to be changed easily)

Reply:
You are correct, all 3 can be based around the vec/slice structs.